File: job/job.go
package job

import (
	"jobmon/config"
	"math"
	"time"
)

const DEFAULT_MAX_POINTS_PER_JOB = 200

type JobMetadata struct {
	Id           int `bun:",pk"`
	UserId       int
	UserName     string
	GroupId      int
	GroupName    string
	ClusterId    string
	NumNodes     int
	NumTasks     int
	TasksPerNode int
	GPUsPerNode  int
	NodeList     string
	StartTime    int
	StopTime     int
	IsRunning    bool
	JobName      string
	Account      string
	TTL          int
	Partition    string
	JobScript    string
	ExitCode     int
	Tags         []*JobTag `bun:"m2m:job_to_tags,join:Job=Tag"`
	Data         []JobMetadataData
}

type JobMetadataData struct {
	Config config.MetricConfig
	Data   float64
}

type StopJob struct {
	ExitCode int
	StopTime int
}

type JobListData struct {
	Jobs   []JobMetadata
	Config JobListConfig
}

type JobListConfig struct {
	Metrics           []string
	RadarChartMetrics []string
	Partitions        map[string]config.PartitionConfig
}

type JobTag struct {
	Id        int64 `bun:",pk,autoincrement"`
	Name      string
	Type      string
	CreatedBy string
}

type JobToTags struct {
	JobId int          `bun:",pk"`
	Job   *JobMetadata `bun:"rel:belongs-to,join:job_id=id"`
	TagId int64        `bun:",pk"`
	Tag   *JobTag      `bun:"rel:belongs-to,join:tag_id=id"`
}

type JobFilter struct {
	UserId    *int
	UserName  *string
	GroupId   *int
	GroupName *string
	IsRunning *bool
	Partition *string
	NumNodes  *RangeFilter
	NumTasks  *RangeFilter
	NumGpus   *RangeFilter
	Time      *RangeFilter
	Tags      *[]JobTag
}

type RangeFilter struct {
	From *int
	To   *int
}

// Check if job TTL has expired.
// If TTL == 0 then the job will never expire
func (j *JobMetadata) Expired() bool {
	now := int(time.Now().Unix())
	if j.TTL == 0 {
		return false
	}
	return j.StopTime+j.TTL < now
}

// Check if job exceeds given maxTime from its start time until now
func (j *JobMetadata) Overtime(maxTime int) bool {
	now := int(time.Now().Unix())
	return j.StartTime+maxTime < now
}

// Calculate the sample interval which displays the closest amount of samples in relation to DEFAULT_MAX_POINTS_PER_JOB
func (j *JobMetadata) CalculateSampleIntervals(metricSampleInterval time.Duration) (intervals []float64, bestAvailableInterval float64) {
	defaultIntervals := []float64{30, 60, 120, 300, 600, 1800}
	duration := float64(j.StopTime - j.StartTime)
	datapoints := duration / metricSampleInterval.Seconds()
	factor := math.Ceil(datapoints / DEFAULT_MAX_POINTS_PER_JOB)
	bestInterval := factor * metricSampleInterval.Seconds()
	bestAvailableInterval = metricSampleInterval.Seconds()
	for _, v := range defaultIntervals {
		if math.Abs(v-bestInterval) < math.Abs(bestAvailableInterval-bestInterval) {
			bestAvailableInterval = v
		}
	}
	for _, v := range defaultIntervals {
		if metricSampleInterval.Seconds() <= v && v <= bestAvailableInterval {
			intervals = append(intervals, v)
		}
	}
	return
}

func (j *JobMetadata) AddTag(tag *JobTag) {
	for _, jt := range j.Tags {
		if jt == tag {
			return
		}
	}
	j.Tags = append(j.Tags, tag)
}

func (j *JobMetadata) RemoveTag(tag *JobTag) {
	var newTags []*JobTag
	for _, jt := range j.Tags {
		if jt != tag {
			newTags = append(newTags, jt)
		}
	}
	j.Tags = newTags
}

func (t *JobTag) IsIn(tags []*JobTag) bool {
	if tags == nil {
		return false
	}
	for _, jt := range tags {
		if jt.Id == t.Id {
			return true
		}
	}
	return false
}

File: frontend/types/job.ts
/**
 * JobMetadata represents all the metadata of a job.
 */
export interface JobMetadata {
  Id: number;
  UserId: number;
  UserName: string;
  GroupId: number;
  GroupName: string;
  ClusterId: string;
  NumNodes: number;
  NumTasks: number;
  TasksPerNode: number;
  GPUsPerNode: number;
  NodeList: string;
  StartTime: number;
  StopTime: number;
  IsRunning: boolean;
  JobName: string;
  Account: string;
  Partition: string;
  JobScript: string;
  Tags: JobTag[];
  Data: JobMetadataData[];
}

/**
 * JobListData stores a list of JobMetadata for a specific configuration.
 */
export interface JobListData {
  Jobs: JobMetadata[];
  Config: JobListConfig;
}

/**
 * JobListConfig stores metrics, partitions, and tags of jobs.
 */
export interface JobListConfig {
  RadarChartMetrics: string[];
  Partitions: DataMap<PartitionConfig>;
  Tags: JobTag[];
}


/**
 * CollectionType gives a list of possible aggregation kinds.
 * @deprecated: not used anywhere.
 */
export enum CollectionType {
  PerNode,
  PerSocket,
  PerThread,
}

/**
 * AggFn give a list of possible aggregation functions.
 */
export enum AggFn {
  Mean = "mean",
  Sum = "sum",
  Min = "min",
  Max = "max",
  Empty = ""
}


/**
 * MetricConfig represents a metric configuration, configured by the admin.
 */
export interface MetricConfig {
  GUID: string;
  Type: string;
  Categories: string[];
  Measurement: string;
  AggFn: AggFn;
  AvailableAggFns: string[];
  SampleInterval: string;
  Unit: string;
  DisplayName: string;
  SeparationKey: string;
  MaxPerNode: number;
  MaxPerType: number;
  PThreadAggFn: AggFn;
  FilterFunc: string;
  PostQueryOp: string;
}

/**
 * MetricPoint represents a metric data point on the visualization chart.
 */
export interface MetricPoint {
  _field: string;
  _measurement: string;
  _start: string;
  _stop: string;
  _time: string;
  _value: number;
  cluster: string;
  hostname: string;
  result: string;
  table: number;
  type: string;
  ["type-id"]: string;
  device: string;
}

/**
 * QuantilePoint represents a a metric data point on the visualization chart 
 * which belongs to a given quantile.
 */
export interface QuantilePoint {
  _field: string;
  _measurement: string;
  _time: string;
  _value: number;
}

/**
 * DataMap 
 */
export interface DataMap<T> {
  [key: string]: T;
}


/**
 * MetricData represents a type of objects used for storing metric data.
 */
export interface MetricData {
  Config: MetricConfig;
  Data: DataMap<MetricPoint[]>;
  RawData: string;
}

/**
 * QuantileData, similar to MetricData except that the data is shown only for 
 * a specific qunatile.
 */
export interface QuantileData {
  Config: MetricConfig;
  Quantiles: string[];
  Data: DataMap<QuantilePoint[]>;
  RawData: string;
}

/**
 * JobData is a generalization of the two previous types, the objects of this 
 * type are used for storing all the necessary data that belonging to jobs.
 */
export interface JobData {
  Metadata: JobMetadata;
  MetricData: MetricData[];
  QuantileData: QuantileData[];
  SampleInterval: number;
  SampleIntervals: number[];
}

/**
 * JobMetaData represents a type of objects used for storing job metadata.
 */
export interface JobMetadataData {
  Config: MetricConfig;
  Mean: number;
  Max: number;
  ChangePoints: string[];
}

/**
 * JobTag represents a type of objects used for storing job tags.
 */
export interface JobTag {
  Id: number;
  Name: string;
  Type: string;
  CreatedBy: string;
}

/**
 * JobSearchParams represents job filters.
 */
export interface JobSearchParams {
  UserId?: number;
  UserName?: string;
  GroupId?: number;
  GroupName?: string;
  IsRunning?: boolean;
  Partition?: string;
  NumNodes?: RangeFilter;
  NumTasks?: RangeFilter;
  NumGpus?: RangeFilter;
  Time?: RangeFilter;
  Tags?: JobTag[];
}

/**
 * RangeFilter represents an integer interval.
 */
type RangeFilter = [number | undefined, number | undefined];


/**
 * VirtualPartitionConfig represents virtual partitions which can be built
 * based on a list of nodes and metrics.
 */
export interface VirtualPartitionConfig {
  Metrics: string[] | null;
  Nodes: string[] | null;
}

/**
 * PartitionConfig represents a partition configuration, which stores
 * the maximum time, a list of metrics and virtual partitions.
 */
export interface PartitionConfig {
  MaxTime: number;
  Metrics: string[] | null;
  VirtualPartitions: DataMap<VirtualPartitionConfig> | null;
}

/**
 * WSMsgType TODO:
 */
export enum WSMsgType {
  LoadMetrics = 1,
  LoadMetricsResponse = 2,
  LatestMetrics = 3,
}

export interface WSMsg {
  Type: number;
}


export interface WSLoadMetricsMsg extends WSMsg {
  StartTime: number;
  StopTime?: number;
}

export interface WSLoadMetricsResponse extends WSMsg {
  MetricData: MetricData[];
}

export type WSLatestMetrics = WSLoadMetricsResponse;
